nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol

```diff
@@ -258,6 +284,106 @@ contract StakingEscrow is Upgradeable, IERC900History {
+        require(
+            _value + getVestedTokens(msg.sender) <= info.value,
+            "Not enough tokens released during vesting"
+        );
+        info.value -= _value;
+
+        token.safeTransfer(msg.sender, _value);
+        emit Withdrawn(msg.sender, _value);
+    }
+
+    /**
+    * @notice Returns amount of not released yet tokens for staker
+    */
+    function getVestedTokens(address _staker) public view returns (uint256) {
+        StakerInfo storage info = stakerInfo[_staker];
+        if (info.vestingReleaseTimestamp <= block.timestamp) {
+            return 0;
+        }
+        uint256 vestedTokens = (info.vestingReleaseTimestamp - block.timestamp) * info.vestingReleaseRate;
+        return info.value < vestedTokens ? info.value : vestedTokens;
+    }
+
+    /**
+    * @notice Setup vesting parameters
+    * @param _stakers Array of stakers
+    * @param _releaseTimestamp Array of timestamps when stake will be released
+    * @param _releaseRate Array of release rates
+    */
+    function setupVesting(
+        address[] calldata _stakers,
+        uint256[] calldata _releaseTimestamp,
+        uint256[] calldata _releaseRate
+    ) external onlyOwner {
+        require(_stakers.length == _releaseTimestamp.length &&
+            _releaseTimestamp.length == _releaseRate.length,
+            "Input arrays must have same number of elements"
+        );
+        for (uint256 i = 0; i < _stakers.length; i++) {
+            address staker = _stakers[i];
+            StakerInfo storage info = stakerInfo[staker];
+            require(info.vestingReleaseTimestamp == 0, "Vesting parameters can be set only once");
+            info.vestingReleaseTimestamp = _releaseTimestamp[i];
+            info.vestingReleaseRate = _releaseRate[i];
+            require(getVestedTokens(staker) > 0, "Vesting parameters must be set properly");
+            emit VestingSet(staker, info.vestingReleaseTimestamp, info.vestingReleaseRate);
+        }
+    }
+
+    /**
+    * @notice Request migration to threshold network
+    * @param _staker Staker address
+    * @param _operator Operator address
+    * @return Amount of tokens
+    */
+    function requestMerge(address _staker, address _operator)
+        external onlyTStakingContract returns (uint256)
+    {
+        StakerInfo storage info = stakerInfo[_staker];
+        require(
+            info.operator == address(0) || info.operator == _operator,
+            "Operator already set for the staker"
+        );
+        if (info.operator == address(0)) {
+            info.operator = _operator;
+            emit MergeRequested(_staker, _operator);
+        }
+        return info.value;
+    }
+
+    /**
+    * @notice Confirm migration to threshold network
+    * @param _staker Staker address
+    */
+    function confirmMerge(address _staker) external {
+        StakerInfo storage info = stakerInfo[_staker];
+        require(info.operator != address(0), "Staker didn't request merge");
```

Comment:
Is "staker didn't request merge" the only outcome of the address is non-zero i.e. is it also possible that the merge already happened, or is this case prevented somewhere before this function is called?

Reply:
`requestMerge` can be called only from T `TokenStaking` during `stakeNu`/`topUpNu`, only these two roots. `requestMerge` is only one entry point so I do not see other way to do it

Reply:
Ah! I misread that line.